examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
-    from.tcp_port = p_tcp.SourcePort;
+#if DEBUG
+        Console.WriteLine("RX {0}:{1} -> {2}:{3} on {4} [{5}]",
+          p_ip.SourceAddress, p_tcp.SourcePort, p_ip.DestinationAddress, p_tcp.DestinationPort, in_port,
+          p_tcp.Syn?"S":"");
+#endif
 
-    int out_port = Port_Drop;
-    if (in_port == Port_Outside)
-    {
-      from.assigned_tcp_port = p_tcp.DestinationPort;
-      from.network_port = null;
-      out_port = outside_to_inside (p_ip, p_tcp, from);
-    } else {
-      from.assigned_tcp_port = null;
-      from.network_port = in_port;
-      out_port = inside_to_outside (p_ip, p_tcp, from);
-    }
+        int out_port = Port_Drop;
+        if (in_port == Port_Outside)
+          out_port = outside_to_inside (p_eth, p_ip, p_tcp);
+        else
+          out_port = inside_to_outside (p_eth, p_ip, p_tcp, in_port);
 
 #if DEBUG
-    print_mapping (port_mapping, " (------ Outside ------) ", " (------ Inside ------) ");
-    Console.WriteLine();
-    print_mapping (port_reverse_mapping, " (------ Inside ------) ", " (------ Outside ------) ");
+        Console.WriteLine(" (------ Outside ------)  \t<->\t (------ Inside ------) ");
+        foreach (var entry in NAT_MapToInside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
+        Console.WriteLine();
+        Console.WriteLine(" (------ Inside ------)  \t<->\t (------ Outside ------) ");
+        foreach (var entry in NAT_MapToOutside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
 #endif
 
-    return out_port;
+        return out_port;
+      }
+    }
+
+    // Drop packets that aren't handled
+    return Port_Drop;
   }
 
-  private int outside_to_inside (IpPacket p_ip, TcpPacket p_tcp, NAT_Entry from)
+  /// <summary>
+  /// Rewrite packets coming from the Outside and forward on the relevant Inside network port.
+  /// </summary>
+  private int outside_to_inside (EthernetPacket p_eth, IpPacket p_ip, TcpPacket p_tcp)
   {
+    // p_ip.DestinationAddress should be my_address
+    if (!p_ip.DestinationAddress.Equals(my_address))
+      return Port_Drop;
+
     // Retrieve the mapping. If a mapping doesn't exist, then it means that we're not
     // aware of a session to which the packet belongs: so drop the packet.
-    NAT_Entry to;
-    if (port_mapping.TryGetValue(from, out to))
+    InternalNode destination;
+    var key = new MapToInside_Key(p_ip.SourceAddress, p_tcp.SourcePort, p_tcp.DestinationPort);
+    if (NAT_MapToInside.TryGetValue(key, out destination))
     {
       // Rewrite destination IP address and TCP port, and map to the appropriate Inside port.
-      p_ip.DestinationAddress = to.ip_address;
-      p_tcp.DestinationPort = to.tcp_port;
-      // Update checksums.
-      p_tcp.UpdateTCPChecksum();
+      p_ip.DestinationAddress = destination.Address;
+      p_tcp.DestinationPort = destination.Port;
+      // Update checksums. NOTE IP updated before TCP
```

Comment:
Order matters when updating checksums at different layers, and it seems intuitive to update from the inside out. Here however it seems important to go against this intuition, and update IP's checksum before TCP's. Did you form an impression why?